From 95dd1825fac8535e34364ec223c8a5ecf6ad40b3 Mon Sep 17 00:00:00 2001
From: Rex Lu <rex.lu@mediatek.com>
Date: Fri, 19 Sep 2025 17:52:26 +0800
Subject: [PATCH 044/105] mtk: mt76: fix RX data queuing of RRO 3.0

For RRO 3.0, RX data should be put to the indicator queue.
Otherwise, "WED OFF + HWRRO ON" won't work.

This patch fixes upstream HWRRO bug, will be upstremed.

Signed-off-by: Rex Lu <rex.lu@mediatek.com>
Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 mac80211.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/mac80211.c b/mac80211.c
index 29d8d0337..f59efec98 100644
--- a/mac80211.c
+++ b/mac80211.c
@@ -875,6 +875,7 @@ static void mt76_rx_release_amsdu(struct mt76_phy *phy, enum mt76_rxq_id q)
 	struct sk_buff *skb = phy->rx_amsdu[q].head;
 	struct mt76_rx_status *status = (struct mt76_rx_status *)skb->cb;
 	struct mt76_dev *dev = phy->dev;
+	struct mt76_queue *rxq = &dev->q_rx[q];
 
 	phy->rx_amsdu[q].head = NULL;
 	phy->rx_amsdu[q].tail = NULL;
@@ -903,6 +904,11 @@ static void mt76_rx_release_amsdu(struct mt76_phy *phy, enum mt76_rxq_id q)
 			return;
 		}
 	}
+
+	/* for RRO 3.0, RX data should be put to the indicator queue */
+	if (mt76_queue_is_wed_rro_data(rxq) && dev->hwrro_mode == MT76_HWRRO_V3)
+		q = MT_RXQ_RRO_IND;
+
 	__skb_queue_tail(&dev->rx_skb[q], skb);
 }
 
-- 
2.45.2

